src/coreclr/jit/optimizer.cpp

```diff
@@ -5431,56 +4844,306 @@ void Compiler::optFindLoops()
     noway_assert(opts.OptimizationEnabled());
     assert(fgDomsComputed);
 
-    optMarkLoopHeads();
+    optMarkLoopHeads();
+
+    // Were there any potential loops in the flow graph?
+
+    if (fgHasLoops)
+    {
+        optFindNaturalLoops();
+    }
+
+    optLoopTableValid = true;
+}
+
+//-----------------------------------------------------------------------------
+// optFindLoopsPhase: The wrapper function for the "find loops" phase.
+//
+PhaseStatus Compiler::optFindLoopsPhase()
+{
+    optFindLoops();
+
+    m_dfsTree = fgComputeDfs();
+    optFindNewLoops();
+
+    if (fgHasLoops)
+    {
+        optFindAndScaleGeneralLoopBlocks();
+    }
+
+    // The old loop table is no longer valid.
+    optLoopTableValid = false;
+    optLoopTable      = nullptr;
+    optLoopCount      = 0;
+
+    // Old dominators and reachability sets are no longer valid.
+    fgDomsComputed = false;
+
+    return PhaseStatus::MODIFIED_EVERYTHING;
+}
+
+//-----------------------------------------------------------------------------
+// optFindNewLoops: Compute new loops and cross validate with old loop table.
+//
+void Compiler::optFindNewLoops()
+{
+    m_loops = FlowGraphNaturalLoops::Find(m_dfsTree);
+
+    if (optCanonicalizeLoops(m_loops))
+    {
+        fgUpdateChangedFlowGraph(FlowGraphUpdates::COMPUTE_DOMS);
+        m_dfsTree = fgComputeDfs();
+        m_loops   = FlowGraphNaturalLoops::Find(m_dfsTree);
+    }
+
+    // Starting now, we require all loops to have pre-headers.
+    optLoopsRequirePreHeaders = true;
+
+    // Leave a bread crumb for future phases like loop alignment about whether
+    // looking for loops makes sense. We generally do not expect phases to
+    // introduce new cycles/loops in the flow graph; if they do, they should
+    // set this to true themselves.
+    // We use more general cycles over "m_loops->NumLoops() > 0" here because
+    // future optimizations can easily cause general cycles to become natural
+    // loops by removing edges.
+    fgMightHaveNaturalLoops = m_dfsTree->HasCycle();
+    assert(fgMightHaveNaturalLoops || (m_loops->NumLoops() == 0));
+}
```

Comment:
I don't know why in the diff this shows up as a big addition... I only added the canonicalization part here.